float32 e1
float32 e2
string color